### ISSUE TITLE
piping

### ISSUE THREAD

**Issue by icambron** [gobblejs/gobble#94]
One thing that frustrates me about the gobble API is that I can't figure out how to compose things. Specifically, I want to do something like this:

```js
var foo = function(node){
  return node.transform('stuff').transform('more-stuff');
}

gobble('src').pipe(foo).moveTo('place');
gobble('otherSrc').pipe(foo).moveTo('otherPlace');
```

I think `pipe` is probably a very simple function on `Node` that looks something like:

```js
pipe(f){
  return f(this);
}
```

So...is there:

 a) a way I'm missing to accomplish this?
 b) a way of hacking this in from userspace (I couldn't figure out how to monkey patch it)?
 c) a chance you'd take that addition as a PR?

**Comment by martypdx** [comment]
You can pass a function to `.transform`, isn't that the same thing?

**Comment by icambron** [comment]
Hmm, but the function passed to transform takes input and output dirs or files, not nodes. It's not obvious to me how to do this sort of thing with `transform` directly.

**Comment by martypdx** [comment]
[File transformers](https://github.com/gobblejs/gobble/wiki/Writing-plugins#file-transformers) take `input, options` where `input` is string contents of file and you return `output`:

```js
gobble( 'src' ).transform( code => '/* bird */' + code ).moveTo( 'place' );
```

**Comment by icambron** [comment]
Right, but I want compose *existing transformations* and reuse those compositions. I don't need to work directly with the strings.

**Comment by msegado** [comment]
This is actually a common enough operation that there's been talk of adding a pipe operator into JS itself: https://github.com/mindeavor/es-pipeline-operator . That's a long way off though, so for now I could see the value of a pipe method for convenience. Biggest downside I see is that it increases the API surface area slightly.

Re. monkey patching, I think you can grab any `Node` _instance_ and add the pipe method to the original prototype as follows (assuming the prototype chain is only one level deep; some experimentation may be needed):
```js
nodeInstance.constructor.prototype.pipe = function pipe(f) { return f(this); };
```

**Comment by icambron** [comment]
Yeah, I played around with that for a bit yesterday and was surprised that I couldn't quite get anything to work. But could easily have been user error.

**Comment by msegado** [comment]
Odd. Give this a shot:
```js
gobble('').constructor.prototype.pipe = function pipe(f) { return f(this); };
```

**Comment by msegado** [comment]
OK... the `Source` and `Merger` nodes returned by `gobble()` are subclasses of `Node`, so we do need to go a level higher in the prototype chain. The following works, albeit with reliance on `__proto__`:
```js
gobble([]).__proto__.__proto__.pipe = function pipe(f) { return f(this); };
```

**Comment by icambron** [comment]
Ah, right, I should have figured that out. Thanks!

**Comment by msegado** [comment]
Np!
